fendermint/testing/contract-test/tests/upgrades.rs

```diff
@@ -0,0 +1,281 @@
+use fendermint_vm_actor_interface::eam::EthAddress;
+use fendermint_vm_core::Timestamp;
+use fendermint_vm_genesis::{Account, Actor, ActorMeta, Genesis, PermissionMode, SignerAddr};
+use fendermint_vm_interpreter::fvm::store::memory::MemoryBlockstore;
+use fendermint_vm_interpreter::fvm::upgrade_scheduler::UpgradeScheduler;
+use fendermint_vm_interpreter::fvm::{
+    bundle::contracts_path, upgrade_scheduler::Upgrade, FvmMessageInterpreter,
+};
+use fendermint_vm_message::chain::ChainMessage;
+
+fn eth_addr_to_h160(eth_addr: &EthAddress) -> H160 {
+    ethers::core::types::Address::from_slice(&eth_addr.0)
+}
+
+// returns a seeded secret key which is guaranteed to be the same every time
+fn my_secret_key() -> SecretKey {
+    SecretKey::random(&mut StdRng::seed_from_u64(123))
+}
+
+// this test applies a series of upgrades to the state and checks that the upgrades are applied correctly
+#[tokio::test]
+async fn test_applying_upgrades() {
+    use bytes::Bytes;
+    use fendermint_rpc::message::{GasParams, MessageFactory};
+    use lazy_static::lazy_static;
+
+    lazy_static! {
+        /// Default gas params based on the testkit.
+        static ref GAS_PARAMS: GasParams = GasParams {
+            gas_limit: 10_000_000_000,
+            gas_fee_cap: TokenAmount::default(),
+            gas_premium: TokenAmount::default(),
+        };
+    }
+
+    // this is the contract we want to deploy
+    const CONTRACT_HEX: &str = include_str!("../../contracts/SimpleCoin.bin");
+    // generate type safe bindings in rust to this contract
+    abigen!(SimpleCoin, "../contracts/SimpleCoin.abi");
+    // once we deploy this contract, this is the address we expect the contract to be deployed to
+    const CONTRACT_ADDRESS: &str = "f410fnz5jdky3zzcj6pejqkomkggw72pcuvkpihz2rwa";
+    // the amount we want to send to the contract
+    const SEND_BALANCE_AMOUNT: u64 = 1000;
+
+    let mut upgrade_scheduler = UpgradeScheduler::new();
+    upgrade_scheduler
+        .add(
+            Upgrade::new("mychain".to_string(), 1, |state| {
+                println!(
+                    "[Upgrade at height {}] Deploy simple contract",
+                    state.block_height()
+                );
+
+                // create a message for deploying the contract
+                let mut mf = MessageFactory::new_secp256k1(my_secret_key(), 1, state.chain_id());
+                let message = match mf
+                    .fevm_create(
+                        Bytes::from(
+                            hex::decode(CONTRACT_HEX)
+                                .context("error parsing contract")
+                                .unwrap(),
+                        ),
+                        Bytes::default(),
+                        TokenAmount::default(),
+                        GAS_PARAMS.clone(),
+                    )
+                    .unwrap()
+                {
+                    ChainMessage::Signed(signed) => signed.into_message(),
```

Comment:
I see, I forgot that the message factory actually signs messages as well, which is a bit inconvenient in our case, to have to produce a throwaway secret key just for this. Maybe there should be an `FvmMessageFactory` underneath that just has the `from` set to some arbitrary FVM `Address` and doesn't sign anything; not sure if it should own and increment the nonce or not. The `MessageFactory` can wrap to sign messages, but we could just not use it here.

Reply:
And then we could deploy as the system actor, although again I forgot how the address would be assigned in that case. Maybe that salted version of actor creation would be useful, with a deterministic salt based on the block height.